registrar/routes_config_watcher_test.go

```diff
@@ -0,0 +1,274 @@
+				Expect(err).NotTo(HaveOccurred())
+				_, err = cfgFile1.Write(routesBytes1)
+				Expect(err).NotTo(HaveOccurred())
+
+				routesBytes2, err := yaml.Marshal(registrar.RoutesConfigSchema{Routes: []config.RouteSchema{route3Schema}})
+				Expect(err).NotTo(HaveOccurred())
+				_, err = cfgFile2.Write(routesBytes2)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("loads all routes from the config", func() {
+				var receivedRoute config.Route
+				Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+				Expect(receivedRoute).To(Equal(route1))
+				Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+				Expect(receivedRoute).To(Equal(route2))
+				Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+				Expect(receivedRoute).To(Equal(route3))
+			})
+
+			Context("when config file is updated and new route is added", func() {
+				It("notifies that route is discovered", func() {
+					var receivedRoute config.Route
+					Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+					Expect(receivedRoute).To(Equal(route1))
+					Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+					Expect(receivedRoute).To(Equal(route2))
+					Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+					Expect(receivedRoute).To(Equal(route3))
+
+					routesBytes2, err := yaml.Marshal(registrar.RoutesConfigSchema{Routes: []config.RouteSchema{route3Schema, route4Schema}})
+					Expect(err).NotTo(HaveOccurred())
+					err = cfgFile2.Truncate(0)
+					Expect(err).NotTo(HaveOccurred())
+					_, err = cfgFile2.Seek(0, 0)
+					Expect(err).NotTo(HaveOccurred())
+					_, err = cfgFile2.Write(routesBytes2)
+					Expect(err).NotTo(HaveOccurred())
+
+					Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+					Expect(receivedRoute).To(Equal(route4))
+				})
+			})
+
+			Context("when config file is updated and route is removed", func() {
+				It("notifies that route is removed", func() {
+					var receivedRoute config.Route
+					Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+					Expect(receivedRoute).To(Equal(route1))
+					Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+					Expect(receivedRoute).To(Equal(route2))
+					Eventually(routesDiscovered, 2).Should(Receive(&receivedRoute))
+					Expect(receivedRoute).To(Equal(route3))
+
+					routesBytes1, err := yaml.Marshal(registrar.RoutesConfigSchema{Routes: []config.RouteSchema{route1Schema}})
+					Expect(err).NotTo(HaveOccurred())
+					err = cfgFile1.Truncate(0)
+					Expect(err).NotTo(HaveOccurred())
+					_, err = cfgFile1.Seek(0, 0)
+					Expect(err).NotTo(HaveOccurred())
+					_, err = cfgFile1.Write(routesBytes1)
+					Expect(err).NotTo(HaveOccurred())
+
+					Eventually(routesRemoved, 2).Should(Receive(&receivedRoute))
+					Expect(receivedRoute).To(Equal(route2))
+
+					Consistently(routesRemoved).ShouldNot(Receive())
```

Comment:
This just means it shouldn't receive after seeing route2 was removed, but not that routes1/route3 haven't been removed while wating for the above eventually to pass, right?

Reply:
No, line 235 will check that the first route that it will get on the channel is route2. And then nothing will be coming on that channel. So that is the only route it gets.

Reply:
BTW, `Eventually(routesRemoved, 2)` means it will wait for 2 seconds, it doesn't wait for 2 items.

Reply:
Ok, i wasn't sure if eventually().Should(receive(thing)) would keep trying when receiving the wrong thing up to the timeout, to ensure it received the right thing. 

eventually().Should(gbytes.Say("message")) works that way by parsing all the logs issued so far, and ignores them if they don't match the message.